client/checkout/woopay/direct-checkout/woopay-direct-checkout.js

```diff
@@ -0,0 +1,254 @@
+		setIsWooPayThirdPartyCookiesEnabled: () => {},
+		setIsUserLoggedIn: () => {},
+		setWoopaySessionResponse: () => {},
+	};
+	/**
+	 * The actionCallback object maps iframe actions (keys) to listeners (values). Take
+	 * "get_is_user_logged_in_success" for example: we send the action "get_is_user_logged_in_success"
+	 * to the WooPayConnectIframe and, when the iframe responds, it uses the "setIsUserLoggedIn" listener.
+	 *
+	 * The reason for this configuration is that it allows us to extend the functionality of the
+	 * WooPayConnectIframe without having to change the WooPayConnectIframe itself.
+	 */
+	static actionCallback = {
+		set_temp_third_party_cookie_success: 'setTempThirdPartyCookie',
+		get_is_third_party_cookies_enabled_success:
+			'setIsWooPayThirdPartyCookiesEnabled',
+		get_is_user_logged_in_success: 'setIsUserLoggedIn',
+		set_redirect_session_data_success: 'setWoopaySessionResponse',
+	};
+
+	static init() {
+		this.state.iframePostMessage = new Promise( ( resolve ) => {
+			this.listeners.setIframePostMessage = resolve;
+		} );
+
+		this.injectWooPayConnectIframe();
+	}
+
+	/**
+	 * Checks if WooPay is enabled.
+	 *
+	 * @return {boolean} True if WooPay is enabled.
+	 */
+	static isWooPayEnabled() {
+		return getConfig( 'isWooPayEnabled' );
+	}
+
+	/**
+	 * Injects the WooPayConnectIframe into the page.
+	 */
+	static injectWooPayConnectIframe() {
+		const appendIframe = ( element ) => {
+			element.parentElement.style.position = 'relative';
+			const hiddenDiv = document.createElement( 'div' );
+			hiddenDiv.style.visibility = 'hidden';
+			hiddenDiv.style.position = 'absolute';
+			hiddenDiv.style.height = '0';
+
+			element.parentElement.appendChild( hiddenDiv );
+
+			ReactDOM.render(
+				<WooPayConnectIframe
+					listeners={ this.listeners }
+					actionCallback={ this.actionCallback }
+				/>,
+				hiddenDiv
+			);
+		};
+
+		const checkoutButtonClassic = document.querySelector(
+			'.wc-proceed-to-checkout .checkout-button'
+		);
+		const checkoutButtonBlocks = document.querySelector(
+			'.wp-block-woocommerce-proceed-to-checkout-block'
+		);
+		if ( checkoutButtonClassic ) {
+			appendIframe( checkoutButtonClassic );
+		} else if ( checkoutButtonBlocks ) {
+			appendIframe( checkoutButtonBlocks );
+		}
+	}
+
+	/**
+	 * Sets a temporary third party cookie in the WooPayConnectIframe.
+	 *
+	 * @return {Promise<*>} Resolves to true if the cookie was set successfully.
+	 */
+	static async setTempThirdPartyCookie() {
+		const setTempThirdPartyCookie = new Promise( ( resolve ) => {
+			this.listeners.setTempThirdPartyCookie = resolve;
+		} );
+
+		const postMessage = await this.state.iframePostMessage;
+		postMessage( { action: 'setTempThirdPartyCookie' } );
+
+		// Since iFrame is refreshed, we need to re-init the iframePostMessage promise.
+		this.state.iframePostMessage = new Promise( ( resolve ) => {
+			this.listeners.setIframePostMessage = resolve;
+		} );
+
+		return await setTempThirdPartyCookie;
```

Comment:
After thinking this over for the weekend I finally realized what was trippping me up most about this piece and it's the `return await`. I'm more used to seeing something like: `await this.setTempThirdPartyCookie()`, at first I thought this might not work because of the listeners in the React component, but with some looking / testing I found that these are already being called as I was thinking in the `direct-checkout/index.js`. Testing locally with removing the `return await`s from these functions everything still seemed to work as expected.

I wouldn't say this is a blocking change, but wanted to mention it to get your opinion and if that also feels cleaner to you (or if there is a piece of the functionality I'm missing that is reliant on the await within these functions)?